projects/RabbitMQ.Client/client/impl/Connection.cs

```diff
@@ -71,6 +70,7 @@ internal sealed class Connection : IConnection
         private volatile bool _running = true;
         private readonly MainSession _session0;
         private SessionManager _sessionManager;
+        internal SemaphoreSlim _writeLock = new SemaphoreSlim(1, 1);
```

Comment:
I haven't measured it but I seem to remember Semaphore is up to 10x slower than a lock. I think I read this once in

http://www.albahari.com/threading/part2.aspx

ah yeah here it is

![image](https://user-images.githubusercontent.com/174258/96266804-5ee66c80-0fc7-11eb-9628-679c5c0cccc5.png)

but these numbers are a bit old and the picture might have changed. I guess as long as we are only doing sync work it might make sense to use the lock instead. The semaphore might come handy when we need to do async work and ensure single writer access

Reply:
Yeah, I'm looking at better solutions, it might be that our previous Channels implementation with the Pipelines just replacing the Stream functionality (for a proper ReadAsync and FlushAsync with minimal allocations) is the best solution. Trying out that approach currently.